File: play/brain.js
/** MaleCNS subgraph dynamics — mirrors c/brain.c (leaky-tanh + MLP readout). */
export class FlyBrain {
  constructor() {
    this.n = 0;
    this.nDn = 0;
    this.nEdges = 0;
    this.rowPtr = null;
    this.colIdx = null;
    this.weight = null;
    this.featureToCell = new Int32Array(6);
    this.visualToCell = [];
    this.dnIndex = new Int32Array(32);
    this.h = null;
    this.params = null;
    this.nParams = 0;
    this.hidden = 16;
    this.nodes = [];
    this.channels = [];
    this.optical = false;
  }

  static async load({ graphUrl, metaUrl, paramsUrl }) {
    const b = new FlyBrain();
    const [graphBuf, meta, paramsBuf] = await Promise.all([
      fetch(graphUrl).then((r) => {
        if (!r.ok) throw new Error(`graph ${r.status}`);
        return r.arrayBuffer();
      }),
      fetch(metaUrl).then((r) => {
        if (!r.ok) throw new Error(`meta ${r.status}`);
        return r.json();
      }),
      fetch(paramsUrl).then((r) => {
        if (!r.ok) throw new Error(`params ${r.status}`);
        return r.arrayBuffer();
      }),
    ]);
    b._loadGraph(graphBuf);
    b.nodes = meta.nodes || [];
    b.channels = meta.channels || [];
    b.visualToCell =
      meta.visual_to_cell ||
      meta.nodes
        .map((n, i) => (n.role === "input" ? i : -1))
        .filter((i) => i >= 0);
    b.setParams(new Float32Array(paramsBuf));
    b.reset();
    return b;
  }

  _loadGraph(buf) {
    const dv = new DataView(buf);
    let o = 0;
    const magic = String.fromCharCode(
      dv.getUint8(0),
      dv.getUint8(1),
      dv.getUint8(2),
      dv.getUint8(3)
    );
    if (magic !== "FFSG") throw new Error("bad FFSG magic");
    o = 4;
    const version = dv.getUint32(o, true); o += 4;
    if (version !== 1) throw new Error(`unsupported FFSG version ${version}`);
    const n = dv.getUint32(o, true); o += 4;
    const nIn = dv.getUint32(o, true); o += 4;
    const nDn = dv.getUint32(o, true); o += 4;
    const nEdges = dv.getUint32(o, true); o += 4;
    if (nIn !== 6) throw new Error("expected 6 inputs");
    this.n = n;
    this.nDn = nDn;
    this.nEdges = nEdges;
    for (let i = 0; i < 6; i++) {
      this.featureToCell[i] = dv.getInt32(o, true);
      o += 4;
    }
    for (let i = 0; i < 32; i++) {
      this.dnIndex[i] = dv.getInt32(o, true);
      o += 4;
    }
    this.rowPtr = new Int32Array(n + 1);
    for (let i = 0; i <= n; i++) {
      this.rowPtr[i] = dv.getInt32(o, true);
      o += 4;
    }
    this.colIdx = new Int32Array(nEdges);
    for (let i = 0; i < nEdges; i++) {
      this.colIdx[i] = dv.getInt32(o, true);
      o += 4;
    }
    this.weight = new Float32Array(nEdges);
    for (let i = 0; i < nEdges; i++) {
      this.weight[i] = dv.getFloat32(o, true);
      o += 4;
    }
    this.h = new Float32Array(n);
    this.nParams = this.hidden * nDn + this.hidden + 2 * this.hidden + 2;
    this.params = new Float32Array(this.nParams);
  }

  setParams(arr) {
    const n = Math.min(arr.length, this.nParams);
    this.params.set(arr.subarray(0, n));
  }

  reset() {
    this.h.fill(0);
  }

  _tanh(x) {
    if (x > 20) return 1;
    if (x < -20) return -1;
    return Math.tanh(x);
  }

  _step(u) {
    const hNew = new Float32Array(this.n);
    for (let i = 0; i < this.n; i++) {
      let acc = u[i];
      const a = this.rowPtr[i];
      const b = this.rowPtr[i + 1];
      for (let k = a; k < b; k++) {
        acc += 1.4 * this.weight[k] * this.h[this.colIdx[k]];
      }
      hNew[i] = 0.3 * this.h[i] + 0.7 * this._tanh(acc);
    }
    this.h.set(hNew);
  }

  _readoutFromState() {
    const p = this.params;
    const nd = this.nDn;
    const H = this.hidden;
    const W1 = 0;
    const b1 = W1 + H * nd;
    const W2 = b1 + H;
    const b2 = W2 + 2 * H;
    const dnActs = new Float32Array(nd);
    for (let i = 0; i < nd; i++) {
      const cell = this.dnIndex[i];
      dnActs[i] = cell >= 0 && cell < this.n ? 4 * this.h[cell] : 0;
    }
    const hidden = new Float32Array(H);
    for (let h = 0; h < H; h++) {
      let a = p[b1 + h];
      for (let i = 0; i < nd; i++) a += p[W1 + h * nd + i] * dnActs[i];
      hidden[h] = this._tanh(a);
    }
    const logits = [0, 0];
    for (let o = 0; o < 2; o++) {
      let a = p[b2 + o];
      for (let h = 0; h < H; h++) a += p[W2 + o * H + h] * hidden[h];
      logits[o] = a;
    }
    const flap = logits[0] > logits[1] ? 1 : 0;
    const m = Math.max(logits[0], logits[1]);
    const e0 = Math.exp(logits[0] - m);
    const e1 = Math.exp(logits[1] - m);
    const pFlap = e0 / (e0 + e1);
    return { flap, logits, pFlap, hidden, dnActs };
  }

  decide(features) {
    const u = new Float32Array(this.n);
    for (let f = 0; f < 6; f++) {
      const cell = this.featureToCell[f];
      if (cell >= 0 && cell < this.n) u[cell] = 2 * (features[f] - 0.5);
    }
    for (let s = 0; s < 3; s++) this._step(u);
    return {
      ...this._readoutFromState(),
      features: Float32Array.from(features),
      drives: null,
    };
  }

  /** Optical path: drives[i] ∈ (-1,1) → visualToCell[i]. */
  decideFromDrives(drives) {
    const u = new Float32Array(this.n);
    const n = Math.min(drives.length, this.visualToCell.length);
    for (let i = 0; i < n; i++) {
      const cell = this.visualToCell[i];
      if (cell >= 0 && cell < this.n) u[cell] = 2 * drives[i];
    }
    for (let s = 0; s < 3; s++) this._step(u);
    return {
      ...this._readoutFromState(),
      features: null,
      drives: Float32Array.from(drives),
    };
  }

  /** Flat readout parameter snapshot for live HUD. */
  paramSnapshot() {
    return Array.from(this.params);
  }
}

/** Draw activity map onto a canvas. */
export function drawBrainMap(canvas, brain, hoverIdx) {
  const ctx = canvas.getContext("2d");
  const n = brain.n;
  const cols = 8;
  const rows = Math.ceil(n / cols);
  const pad = 4;
  const cell = Math.floor(
    Math.min(
      (canvas.width - pad * 2) / cols,
      (canvas.height - pad * 2 - 28) / rows
    )
  );
  ctx.fillStyle = "#0b1020";
  ctx.fillRect(0, 0, canvas.width, canvas.height);
  ctx.fillStyle = "#9ab";
  ctx.font = "11px ui-monospace, Menlo, monospace";
  ctx.fillText("MaleCNS subgraph activity", pad, 14);

  const dnSet = new Set();
  for (let i = 0; i < brain.nDn; i++) dnSet.add(brain.dnIndex[i]);
  const inSet = new Set(
    brain.visualToCell && brain.visualToCell.length
      ? brain.visualToCell
      : [...brain.featureToCell]
  );

  for (let i = 0; i < n; i++) {
    const c = i % cols;
    const r = Math.floor(i / cols);
    const x = pad + c * cell;
    const y = 22 + pad + r * cell;
    const v = brain.h[i];
    const t = Math.max(-1, Math.min(1, v));
    // blue ← 0 → amber
    if (t >= 0) {
      const g = Math.floor(40 + t * 180);
      const rC = Math.floor(40 + t * 215);
      ctx.fillStyle = `rgb(${rC},${g},40)`;
    } else {
      const b = Math.floor(40 + -t * 200);
      ctx.fillStyle = `rgb(30,60,${b})`;
    }
    ctx.fillRect(x + 1, y + 1, cell - 2, cell - 2);

    let stroke = "#334";
    if (inSet.has(i)) stroke = "#4ade80";
    if (dnSet.has(i)) stroke = "#f472b6";
    ctx.strokeStyle = i === hoverIdx ? "#fff" : stroke;
    ctx.lineWidth = i === hoverIdx ? 2 : 1;
    ctx.strokeRect(x + 1, y + 1, cell - 2, cell - 2);
  }

  // legend
  const ly = canvas.height - 14;
  ctx.fillStyle = "#4ade80";
  ctx.fillRect(pad, ly - 8, 8, 8);
  ctx.fillStyle = "#9ab";
  ctx.fillText("input", pad + 12, ly);
  ctx.fillStyle = "#f472b6";
  ctx.fillRect(pad + 60, ly - 8, 8, 8);
  ctx.fillStyle = "#9ab";
  ctx.fillText("DN out", pad + 72, ly);
}

File: c/brain.c
#include "brain.h"

#include <math.h>
#include <stdio.h>
#include <string.h>

static float tanhf_local(float x) {
  if (x > 20.0f)
    return 1.0f;
  if (x < -20.0f)
    return -1.0f;
  return tanhf(x);
}

static void brain_alloc_params(FfBrain *b) {
  b->n_params = FF_BRAIN_HIDDEN * b->n_dn + FF_BRAIN_HIDDEN +
                2 * FF_BRAIN_HIDDEN + 2;
  if (b->n_params > (int)(sizeof(b->params) / sizeof(b->params[0])))
    b->n_params = (int)(sizeof(b->params) / sizeof(b->params[0]));
  for (int i = 0; i < b->n_params; i++)
    b->params[i] = 0.0f;
}

void brain_init_identity_stub(FfBrain *b) {
  memset(b, 0, sizeof(*b));
  b->n_dn = FF_N_FEATURES;
  b->n = FF_N_FEATURES + FF_N_FEATURES;
  for (int i = 0; i < FF_N_FEATURES; i++) {
    b->feature_to_cell[i] = i;
    b->dn_index[i] = FF_N_FEATURES + i;
  }
  int e = 0;
  for (int post = 0; post < b->n; post++) {
    b->row_ptr[post] = e;
    if (post >= FF_N_FEATURES) {
      b->col_idx[e] = post - FF_N_FEATURES;
      b->weight[e] = 1.0f;
      e++;
    }
  }
  b->row_ptr[b->n] = e;
  b->n_edges = e;
  brain_alloc_params(b);
}

int brain_load_bin(FfBrain *b, const char *path) {
  memset(b, 0, sizeof(*b));
  FILE *f = fopen(path, "rb");
  if (!f)
    return -1;
  char magic[4];
  if (fread(magic, 1, 4, f) != 4 || memcmp(magic, "FFSG", 4) != 0) {
    fclose(f);
    return -2;
  }
  unsigned version = 0, n = 0, n_in = 0, n_dn = 0, n_edges = 0;
  if (fread(&version, 4, 1, f) != 1 || fread(&n, 4, 1, f) != 1 ||
      fread(&n_in, 4, 1, f) != 1 || fread(&n_dn, 4, 1, f) != 1 ||
      fread(&n_edges, 4, 1, f) != 1) {
    fclose(f);
    return -3;
  }
  if (n > FF_BRAIN_MAX_N || n_edges > FF_BRAIN_MAX_E || n_dn > FF_BRAIN_MAX_DN ||
      n_in != FF_N_FEATURES) {
    fclose(f);
    return -4;
  }
  b->n = (int)n;
  b->n_dn = (int)n_dn;
  b->n_edges = (int)n_edges;
  if (fread(b->feature_to_cell, 4, FF_N_FEATURES, f) != FF_N_FEATURES) {
    fclose(f);
    return -5;
  }
  int dn_pad[FF_BRAIN_MAX_DN];
  if (fread(dn_pad, 4, FF_BRAIN_MAX_DN, f) != FF_BRAIN_MAX_DN) {
    fclose(f);
    return -6;
  }
  memcpy(b->dn_index, dn_pad, sizeof(int) * b->n_dn);
  if (fread(b->row_ptr, 4, n + 1, f) != n + 1) {
    fclose(f);
    return -7;
  }
  if (fread(b->col_idx, 4, n_edges, f) != n_edges) {
    fclose(f);
    return -8;
  }
  if (fread(b->weight, 4, n_edges, f) != n_edges) {
    fclose(f);
    return -9;
  }
  fclose(f);
  brain_alloc_params(b);
  return 0;
}

void brain_reset(FfBrain *b) {
  for (int i = 0; i < b->n; i++)
    b->h[i] = 0.0f;
}

void brain_set_silenced(FfBrain *b, bool on) { b->silenced = on; }

int brain_n_params(const FfBrain *b) { return b->n_params; }

void brain_set_params(FfBrain *b, const float *params, int n) {
  if (n > b->n_params)
    n = b->n_params;
  memcpy(b->params, params, (size_t)n * sizeof(float));
}

static void recurrent_step(FfBrain *b, const float *u) {
  float h_new[FF_BRAIN_MAX_N];
  for (int i = 0; i < b->n; i++) {
    float acc = u[i];
    for (int k = b->row_ptr[i]; k < b->row_ptr[i + 1]; k++) {
      int j = b->col_idx[k];
      acc += 1.4f * b->weight[k] * b->h[j];
    }
    h_new[i] = 0.3f * b->h[i] + 0.7f * tanhf_local(acc);
  }
  memcpy(b->h, h_new, (size_t)b->n * sizeof(float));
}

static void readout_logits(const FfBrain *b, float logits[2]) {
  const float *p = b->params;
  int nd = b->n_dn;
  const float *W1 = p;
  const float *b1 = W1 + FF_BRAIN_HIDDEN * nd;
  const float *W2 = b1 + FF_BRAIN_HIDDEN;
  const float *b2 = W2 + 2 * FF_BRAIN_HIDDEN;

  float hidden[FF_BRAIN_HIDDEN];
  for (int h = 0; h < FF_BRAIN_HIDDEN; h++) {
    float a = b1[h];
    for (int i = 0; i < nd; i++) {
      float act = 0.0f;
      if (!b->silenced) {
        int cell = b->dn_index[i];
        if (cell >= 0 && cell < b->n)
          act = 4.0f * b->h[cell];
      }
      a += W1[h * nd + i] * act;
    }
    hidden[h] = tanhf_local(a);
  }
  for (int o = 0; o < 2; o++) {
    float a = b2[o];
    for (int h = 0; h < FF_BRAIN_HIDDEN; h++)
      a += W2[o * FF_BRAIN_HIDDEN + h] * hidden[h];
    logits[o] = a;
  }
}

int brain_decide(FfBrain *b, const float features[FF_N_FEATURES]) {
  float u[FF_BRAIN_MAX_N];
  for (int i = 0; i < b->n; i++)
    u[i] = 0.0f;
  if (!b->silenced) {
    for (int f = 0; f < FF_N_FEATURES; f++) {
      int cell = b->feature_to_cell[f];
      if (cell >= 0 && cell < b->n)
        u[cell] = 2.0f * (features[f] - 0.5f);
    }
  }
  for (int s = 0; s < 3; s++)
    recurrent_step(b, u);

  float logits[2];
  readout_logits(b, logits);
  return logits[0] > logits[1] ? 1 : 0;
}
